cpp/include/cudf/detail/aggregation/aggregation.cuh

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/aggregation.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/table/table_device_view.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+template <typename Source, aggregation::Kind k, typename Enable = void>
+struct update_target_element {};
+
+template <typename Source, aggregation::Kind k>
+struct update_target_element<
+    Source, k, std::enable_if_t<is_valid_aggregation<Source, k>()>> {};
+
+template <typename Source>
+struct update_target_element<
+    Source, aggregation::COUNT,
+    std::enable_if_t<is_valid_aggregation<Source, aggregation::COUNT>()>> {};
+
+template <typename Source>
+struct element_wise_aggregator_impl {
+  template <aggregation::Kind k>
+  __device__ inline void operator()(mutable_column_device_view target,
+                                    size_type target_index,
+                                    column_device_view source,
+                                    size_type source_index) {}
+};
+
+struct elementwise_aggregator {
+  template <typename Source>
+  __device__ inline void operator()(mutable_column_device_view target,
```

Comment:
I see no reason why `operator()(...)` shouldn't be `const` (since `elementwise_aggregator` is stateless, hence immutable by any of its memfs)